arraycontext/impl/pyopencl/taggable_cl_array.py

```diff
@@ -0,0 +1,127 @@
+                                 " at once.")
+            tags = self.tags if tags is None else tags
+            axes = self.axes if axes is None else axes
+            return self.__class__(None, self.shape, self.dtype,
+                                  allocator=self.allocator,
+                                  strides=self.strides, data=self.base_data,
+                                  offset=self.offset, events=self.events,
+                                  _fast=True, _context=self.context,
+                                  _queue=self.queue, _size=self.size,
+                                  tags=tags, axes=axes)
+        else:
+            new_with_queue = super().copy(queue=queue)
+            return self.__class__(None, new_with_queue.shape,
+                                  new_with_queue.dtype,
+                                  allocator=new_with_queue.allocator,
+                                  strides=new_with_queue.strides,
+                                  data=new_with_queue.base_data,
+                                  offset=new_with_queue.offset,
+                                  events=new_with_queue.events, _fast=True,
+                                  _context=new_with_queue.context,
+                                  _queue=new_with_queue.queue,
+                                  _size=new_with_queue.size,
+                                  tags=self.tags,
+                                  axes=self.axes,
+                                  )
+
+    def with_tagged_axis(self, iaxis: int,
+                         tags: Union[Sequence[Tag], Tag]) -> "TaggableCLArray":
+        """
+        Returns a copy of *self* with *iaxis*-th axis tagged with *tags*.
+        """
+        new_axes = (self.axes[:iaxis]
+                    + (self.axes[iaxis].tagged(tags),)
+                    + self.axes[iaxis+1:])
+        return self.copy(axes=new_axes)
+
+
+def to_tagged_cl_array(ary: cla.Array,
+                       axes: Optional[Tuple[Axis, ...]],
+                       tags: FrozenSet[Tag]) -> TaggableCLArray:
+    """
+    Returns a :class:`TaggableCLArray` that is constructed from the data in
+    *ary* along with the metadata from *axes* and *tags*.
+
+    :arg axes: An instance of :class:`Axis` for each dimension of the
+        array. If passed *None*, then initialized to a :class:`pytato.Axis`
+        with no tags attached for each dimension.
+    """
+    return TaggableCLArray(None, ary.shape,
+                           ary.dtype,
+                           allocator=ary.allocator,
+                           strides=ary.strides,
+                           data=ary.base_data,
+                           offset=ary.offset,
+                           events=ary.events, _fast=True,
+                           _context=ary.context,
+                           _queue=ary.queue, _size=ary.size,
+                           axes=axes,
+                           tags=tags)
```

Comment:
I don't love the idea of hardcoding so much of the internal structure of `cla.Array` here. Is there a copy method in `Array` that we could use instead?

Reply:
Not quite, see [cla.Array.copy](https://github.com/inducer/pyopencl/blob/fb13a7dfe4c6c07b0661fb554663bc403c28b97c/pyopencl/array.py#L828). But tried to avoid the duplication by invoking `TaggableCLArray.copy`. So there's only one source location where we rely on the structure on `cl.array.Array.__init__`'s signature.